[Enhance] Add param to control save checkpoint

zhijiejia opened the pull request:
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

Model training, sometimes do not need to save the previous weight, just need a later epoch to start saving some weights, and `max_keep_ckpts` still need to save the previous weight, but save_start will not save the previous weight

## Modification

add param `save_start` in `mmengine/hooks/checkpoint_hook.py`  `Class CheckpointHook`

## BC-breaking (Optional)

## Use cases (Optional)

## Checklist

HAOCHENYE commented on mmengine/hooks/checkpoint_hook.py:
```diff
@@ -55,6 +55,8 @@ class CheckpointHook(Hook):
             detection and instance segmentation. ``AR@100`` for proposal
             recall. If ``save_best`` is ``auto``, the first key of the returned
             ``OrderedDict`` result will be used. Defaults to None.
+        save_start (int): Start saving the number of epochs or
+            iterations of model weights. Defaults to 0.
```
`save_start` should be added as the last argument to prevent from BC breaking.

zhijiejia replied:
Ok, by the way, if you put the save_start parameter at the end, won't it look uncomfortable? Because it is not put together with the previous save_ * parameter? I found that `mmsegmentation` and `mmdetection` also use `key-val` to pass parameters. Does this mean that you don't need to put save_ * at the end?

zhijiejia replied:
Thank you for your late night reply.

HAOCHENYE commented on mmengine/hooks/checkpoint_hook.py:
```diff
@@ -316,6 +320,10 @@ def _save_checkpoint(self, runner) -> None:
         Args:
             runner (Runner): The runner of the training process.
         """
+        if (self.by_epoch and runner.epoch < self.save_start) or (
+                not self.by_epoch and runner.iter < self.save_start):
+            return
```
should this code be added to `after_train_iter` and `after_train_epoch`?

zhijiejia replied:
I found that `after_train_iter` function is placed at the end of the `checkpoint_hook.py`, why not put it after the `after_val_epoch`, at least they have similar functions

zhijiejia replied:
In my opinion, the `every_n_train_iters` and `every_n_epochs` functions also need to be rewritten, because with `save_start` you should think `save_start` with `self.interval`

HAOCHENYE replied:
> I found that `after_train_iter` function is placed at the end of the `checkpoint_hook.py`, why not put it after the `after_val_epoch`, at least they have similar functions

Agree :rofl:, I think `after_train_iter` is not placed in the correct position either. 

I think we should keep the single responsibility of `every_n_iter` and `every_n_epoch`.  Their functions should be the same as their name expressed. If we want to add an extra check, like `save_start`, will the following way more proper?

```python
    def after_train_iter(self,
                         runner,
                         batch_idx: int,
                         data_batch: DATA_BATCH = None,
                         outputs=Optional[dict]) -> None:
        if self.by_epoch or runner.iter < self.save_start:
            return
        # save checkpoint for following cases:
        # 1. every ``self.interval`` iterations
        # 2. reach the last iteration of training
        if self.every_n_train_iters(runner, self.interval) or \
                (self.save_last and
                 self.is_last_train_iter(runner)):
            runner.logger.info(
                f'Saving checkpoint at {runner.iter + 1} iterations')
            self._save_checkpoint(runner)
```

zhijiejia replied:
But if you set `save_start` and `interval` at the same time, then it should start from `save_start` and save the weights once every `interval`? So I would think that `every_n_train_iters` and `every_n_epochs` code should be modified to meet `save_start` scenarios

HAOCHENYE replied:
Got it, it makes sense to me.

HAOCHENYE reviewed (commented):
Thanks for your contributions, it could be better to add the unit test for `save_start`

HAOCHENYE reviewed (commented):
LGTM, and It could be better to add a corresponding unit test to 
https://github.com/open-mmlab/mmengine/blob/601db12d38245140775ed189fcc3b4815843b567/tests/test_hooks/test_checkpoint_hook.py#L356

and 

https://github.com/open-mmlab/mmengine/blob/601db12d38245140775ed189fcc3b4815843b567/tests/test_hooks/test_checkpoint_hook.py#L423

zhijiejia commented on mmengine/hooks/checkpoint_hook.py:
```diff
@@ -316,6 +323,7 @@ def _save_checkpoint(self, runner) -> None:
         Args:
             runner (Runner): The runner of the training process.
         """
+
```
ok， i will study for add it

zhouzaida commented on mmengine/hooks/checkpoint_hook.py:
```diff
@@ -84,6 +84,8 @@ class CheckpointHook(Hook):
             preifx of uri corresponding backend. Defaults to None.
             New in v0.2.0.
+        save_start (int): Start saving the number of epochs or
+            iterations of model weights. Defaults to 0.
```
```suggestion
            iterations of model weights. Defaults to 0.
            New in v0.3.0.
```

zhijiejia replied:
ok， thanks for your help

HAOCHENYE commented:
Hi, it could be better to add a unit test to test `save_start` works correctly.

zhijiejia commented:
> Hi, it could be better to add a unit test to test `save_start` works correctly.

yes, please give me some time, I'm preparing my homework, and when it's done, I can continue to add unit tests...

zhijiejia commented:
> > Hi, it could be better to add a unit test to test `save_start` works correctly.
> 
> yes, please give me some time, I'm preparing my homework, and when it's done, I can continue to add unit tests...
Friday or Saturday I will add the test code.

HAOCHENYE commented:
> > > > Hi, it could be better to add a unit test to test `save_start` works correctly.
> > > 
> > > 
> > > yes, please give me some time, I'm preparing my homework, and when it's done, I can continue to add unit tests...
> > 
> > 
> > Friday or Saturday I will add the test code.
> 
> Please, I added `save_start` parameter to `test_with_runner` in `test_checkpoint_hook.py` and passed it to `checkpoint_cfg`, so is this a test for that `save_start` parameter ?

Yeah, we should test that the name of the saved checkpoint starts with `epoch_{save_start}`

zhijiejia commented:
thanks, I've been obsessed with `epoch + 1`, I don't know why `(runner.epoch + 1) % n`  why not `runner.epoch % n` https://github.com/open-mmlab/mmengine/blob/64ac14303fc786151beabba1de4ddfd7c2e5a408/mmengine/hooks/hook.py#L343

zhijiejia commented:
yes，the code in `test_with_runner` should be change, I thought this loop code could not be changed, it should be `max_epoch` - > `max_epoch + 1`

zhijiejia commented:
> > > > > Hi, it could be better to add a unit test to test `save_start` works correctly.
> > > > 
> > > > 
> > > > yes, please give me some time, I'm preparing my homework, and when it's done, I can continue to add unit tests...
> > > 
> > > 
> > > Friday or Saturday I will add the test code.
> > 
> > 
> > Please, I added `save_start` parameter to `test_with_runner` in `test_checkpoint_hook.py` and passed it to `checkpoint_cfg`, so is this a test for that `save_start` parameter ?
> 
> Yeah, we should also test that the name of the first saved checkpoint namedwith `epoch_{save_start}`

Please, Is the unit test completed now?

HAOCHENYE commented:
> > > > > > Hi, it could be better to add a unit test to test `save_start` works correctly.
> > > > > 
> > > > > 
> > > > > yes, please give me some time, I'm preparing my homework, and when it's done, I can continue to add unit tests...
> > > > 
> > > > 
> > > > Friday or Saturday I will add the test code.
> > > 
> > > 
> > > Please, I added `save_start` parameter to `test_with_runner` in `test_checkpoint_hook.py` and passed it to `checkpoint_cfg`, so is this a test for that `save_start` parameter ?
> > 
> > 
> > Yeah, we should also test that the name of the first saved checkpoint namedwith `epoch_{save_start}`
> 
> Please, Is the unit test completed now?

Yeah, the original unit test seems a bit weird, so I modified `max_epochs` to `max_epochs+1` 🤣.  Hmmm, I think the unit test has covered this PR, do you have any supplementation?

BTW, the `epoch+1` and `iter+1` also make me very uncomfortable 🤣, I'll always spend some time to check should I `+1` or not..

zhijiejia commented:
> > > > > > > Hi, it could be better to add a unit test to test `save_start` works correctly.
> > > > > > 
> > > > > > 
> > > > > > yes, please give me some time, I'm preparing my homework, and when it's done, I can continue to add unit tests...
> > > > > 
> > > > > 
> > > > > Friday or Saturday I will add the test code.
> > > > 
> > > > 
> > > > Please, I added `save_start` parameter to `test_with_runner` in `test_checkpoint_hook.py` and passed it to `checkpoint_cfg`, so is this a test for that `save_start` parameter ?
> > > 
> > > 
> > > Yeah, we should also test that the name of the first saved checkpoint namedwith `epoch_{save_start}`
> > 
> > 
> > Please, Is the unit test completed now?
> 
> Yeah, the original unit test seems a bit weird, so I modified `max_epochs` to `max_epochs+1` 🤣. Hmmm, I think the unit test has covered this PR, do you have any supplementation?
> 
> BTW, the `epoch+1` and `iter+1` also make me very uncomfortable 🤣, I'll always spend some time to check should I `+1` or not..

As mentioned earlier, the position adjustment of that function
https://github.com/open-mmlab/mmengine/blob/main/mmengine/hooks/checkpoint_hook.py#L500

zhouzaida commented on mmengine/hooks/checkpoint_hook.py:
```diff
@@ -84,6 +84,8 @@ class CheckpointHook(Hook):
             preifx of uri corresponding backend. Defaults to None.
             New in v0.2.0.
+        save_start (int): Start saving the number of epochs or
+            iterations of model weights. Defaults to 0.
```
```suggestion
            iterations of model weights. Defaults to 0.
            New in version 0.3.0.
```

zhouzaida commented on mmengine/hooks/checkpoint_hook.py:
```diff
@@ -128,6 +130,7 @@ def __init__(self,
                  file_client_args: Optional[dict] = None,
                  filename_tmpl: Optional[str] = None,
                  backend_args: Optional[dict] = None,
+                 save_start: int = 0,
```
suggest renaming `save_start` to `begin`.

zhijiejia replied:
emm, I think `save_start` more in line with what you see is what you get, more in line with what it means to start saving weight?

zhouzaida replied:
Will `begin` be ambiguous? The function of this hook is to save checkpoints so `begin` should mean starting saving checkpoints.

zhouzaida replied:
Okay. Can we rename it to `save_begin` which is consistent with `val_begin` at `mmengine/runner/loops.py`.

https://github.com/open-mmlab/mmengine/blob/4e685931d0687b45323b9811501490bfde2a6e11/mmengine/runner/loops.py#L26.

zhijiejia replied:
ok，i agree with you

zhouzaida commented on mmengine/hooks/checkpoint_hook.py:
```diff
@@ -84,6 +84,9 @@ class CheckpointHook(Hook):
         backend_args (dict, optional): Arguments to instantiate the
             preifx of uri corresponding backend. Defaults to None.
             New in v0.2.0.
+        save_begin (int): Start saving the number of epochs or
```
It may be more intuitive to default to 1.

HAOCHENYE replied:
```suggestion
        save_begin (int): The number of iterations or epochs when starting to save the checkpoint
```

The last comment :smile:

zhouzaida commented on mmengine/hooks/checkpoint_hook.py:
```diff
@@ -84,6 +84,9 @@ class CheckpointHook(Hook):
             New in v0.2.0.
+        save_begin (int): Start saving the number of epochs or
+            epochs or iterations of model weights. Defaults to 1.
+            New in v0.3.0.
```
```suggestion
            New in v0.4.0.
```

zhouzaida commented on mmengine/hooks/checkpoint_hook.py:
```diff
@@ -316,6 +354,7 @@ def _save_checkpoint(self, runner) -> None:
         Args:
             runner (Runner): The runner of the training process.
         """
+
```
```suggestion
```

zhijiejia commented:
So, can this PR be merged? Can you give some suggestions? I said above that the `save_begin` must start from `1` by `assert`. Do you agree? Because I haven't answered my question for several days, I don't know what the situation is.

zhouzaida commented:
> So, can this PR be merged? Can you give some suggestions? I said above that the `save_begin` must start from `1` by `assert`. Do you agree? Because I haven't answered my question for several days, I don't know what the situation is.

This PR will be merged today or tomorrow.

zhijiejia commented:
So, are there any next suggestions?

HAOCHENYE reviewed (dismissed):
Approved!!!
![image](https://user-images.githubusercontent.com/57566630/200800309-6c3f6f28-845c-4e1b-b437-d3087e9aa7a9.png)

zhijiejia commented:
Thanks for your help, so can we do next (merge haha) ?

zhouzaida commented:
Hi @zhijiejia , lint failed.

zhouzaida commented on mmengine/hooks/checkpoint_hook.py:
```diff
@@ -84,6 +84,9 @@ class CheckpointHook(Hook):
         backend_args (dict, optional): Arguments to instantiate the
             preifx of uri corresponding backend. Defaults to None.
             New in v0.2.0.
+        save_begin (int): The number of iterations or epochs when starting to save the checkpoint
+            epochs or iterations of model weights. Defaults to 1.
+            New in v0.4.0.
```
```suggestion
        save_begin (int): The number of iterations or epochs when starting to
            save the checkpoint. Defaults to 1.
            New in v0.4.0.
```

zhijiejia commented:
thanks, I saw it in the cafeteria and then got busy after dinner and forgot about it...

zhouzaida commented on mmengine/hooks/checkpoint_hook.py:
```diff
@@ -286,6 +294,36 @@ def after_train_epoch(self, runner) -> None:
+        # save checkpoint for following cases:
+        # 1. every ``self.interval`` iterations
+        # 2. reach the last iteration of training
+        if self.every_n_train_iters(runner, self.interval) or \
```
When `save_begin` is equal to 1, the result of every_n_train_iters is not the same as before adding this parameter.

- Before

```python
def every_n_train_iters(self, runner, n: int) -> bool:
    return (runner.iter + 1) % n == 0 if n > 0 else False
```

- Adding this parameter

```python
def every_n_train_iters(self, runner, n: int) -> bool:
    return (runner.iter + 1 - 1) % n == 0 if n > 0 else False
```

HAOCHENYE replied:
Why we do not override `every_n_train_iter` anymore? It seems the base `every_n_train_iter` will not consider the `save_begin`

zhouzaida commented:
Hi @zhijiejia , is there any progress?

zhijiejia commented:
> Hi @zhijiejia , is there any progress?

yes, i

> Hi @zhijiejia , is there any progress?

yes, I try to modify another logic, not to modify the logic of `every_n_train_iters`

zhijiejia commented:
The point I modified is that I just added `save_begin` parameters to determine whether it can be saved when the weight was originally saved

zhijiejia commented:
All test case in `tests/test_hooks/test_checkpoint_hook.py` is passed

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/open-mmlab/mmengine?pullRequest=636) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/open-mmlab/mmengine?pullRequest=636) before we can accept your contribution.<br/>**2** out of **3** committers have signed the CLA.<br/><br/>:white_check_mark: HAOCHENYE<br/>:white_check_mark: zhouzaida<br/>:x: zhijiejia<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/open-mmlab/mmengine?pullRequest=636) it.</sub>

HAOCHENYE commented:
Hi, is there any progress? CLA is out of date, please sign the CLA again :smile:
